libraries/previousAuctionInfo/previousAuctionInfo.js

```diff
@@ -0,0 +1,130 @@
+import {on as onEvent} from '../../src/events.js';
+import { EVENTS } from '../../src/constants.js';
+
+export let previousAuctionInfoEnabled = false;
+let enabledBidders = [];
+export let winningBidsMap = {};
+
+export const resetPreviousAuctionInfo = () => {
+  previousAuctionInfoEnabled = false;
+  enabledBidders = [];
+  winningBidsMap = {};
+};
+
+export const enablePreviousAuctionInfo = (config, cb = initHandlers) => {
+  const { bidderCode, isBidRequestValid } = config;
+  const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderCode);
+  if (!enabledBidder) enabledBidders.push({ bidderCode, isBidRequestValid, maxQueueLength: config.maxQueueLength || 10 });
+  if (previousAuctionInfoEnabled) return;
+  previousAuctionInfoEnabled = true;
+  cb();
+}
+
+export const initHandlers = () => {
+  onEvent(EVENTS.AUCTION_END, onAuctionEndHandler);
+  onEvent(EVENTS.BID_WON, onBidWonHandler);
+  onEvent(EVENTS.BID_REQUESTED, onBidRequestedHandler);
+};
+
+export const onAuctionEndHandler = (auctionDetails) => {
+  // eslint-disable-next-line no-console
+  console.log('onAuctionEndHandler', auctionDetails);
+
+  try {
+    let highestCpmBid = 0;
+    const receivedBidsMap = {};
+    const rejectedBidsMap = {};
+
+    if (auctionDetails.bidsReceived && auctionDetails.bidsReceived.length) {
+      highestCpmBid = auctionDetails.bidsReceived.reduce((highestBid, currentBid) => {
+        return currentBid.cpm > highestBid.cpm ? currentBid : highestBid;
+      }, auctionDetails.bidsReceived[0]);
+
+      auctionDetails.bidsReceived.forEach(bidReceived => {
+        receivedBidsMap[bidReceived.requestId] = bidReceived;
+      });
+    }
+
+    if (auctionDetails.bidsRejected && auctionDetails.bidsRejected.length) {
+      auctionDetails.bidsRejected.forEach(bidRejected => {
+        rejectedBidsMap[bidRejected.requestId] = bidRejected;
+      });
+    }
+
+    if (auctionDetails.bidderRequests && auctionDetails.bidderRequests.length) {
+      auctionDetails.bidderRequests.forEach(bidderRequest => {
+        const enabledBidder = enabledBidders.find(bidder => bidder.bidderCode === bidderRequest.bidderCode);
+
+        if (enabledBidder) {
+          bidderRequest.bids.forEach(bid => {
+            const isValidBid = enabledBidder.isBidRequestValid(bid);
+
+            if (!isValidBid) return;
+
+            const previousAuctionInfoPayload = {
+              bidderRequestId: bidderRequest.bidderRequestId,
+              minBidToWin: highestCpmBid?.cpm || 0,
```

Comment:
How does 'minBidToWin' differ from the 'highestcpm' field? I can live with this term because it's sort of an industry standard, but it's actually not correct. Not guaranteed at all that they would have won with this bid. It just happens to be the highest bid that Prebid knows about. Which is why I prefer the term 'highestcpm' -- there's less of a promise involved. But anyhow, I get that 'minBidToWin' is a cow that's left the barn, so I would suggest removing 'highestcpm' then. Please make sure that it's positioned right next to the `cur` field though. Another pet peeve of mine is when it's not clear what currency a price is stated in.

Which I suppose opens a can of worms. The obvious minBidToWin currency is the ortb.cur, but there are going to bidders that will only accept minBidToWin in USD, EUR, or whatever. Do we need this module to convert to different currencies, so an array of minBidToWin entries?